session/portforward.go

```diff
@@ -0,0 +1,335 @@
+}
+
+func (f *MenderPortForwarder) Connect(protocol string, host string, portNumber uint) error {
+	log.Debugf("port-forward[%s/%s] connect: %s/%s:%d", f.SessionID, f.ConnectionID, protocol, host, portNumber)
+
+	if protocol == wspf.PortForwardProtocolTCP || protocol == wspf.PortForwardProtocolUDP {
+		conn, err := net.Dial(protocol, host+":"+strconv.Itoa(int(portNumber)))
+		if err != nil {
+			return err
+		}
+		f.conn = conn
+	} else {
+		return errors.New("unknown protocol: " + protocol)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	f.ctx = ctx
+	f.ctxCancel = cancelFunc
+
+	go f.Read()
+
+	return nil
+}
+
+func (f *MenderPortForwarder) Close(sendStopMessage bool) error {
+	if f.closed {
+		return nil
+	}
+	f.closed = true
+	log.Debugf("port-forward[%s/%s] close", f.SessionID, f.ConnectionID)
+	if sendStopMessage {
+		m := &ws.ProtoMsg{
+			Header: ws.ProtoHdr{
+				Proto:     ws.ProtoTypePortForward,
+				MsgType:   wspf.MessageTypePortForwardStop,
+				SessionID: f.SessionID,
+				Properties: map[string]interface{}{
+					propertyConnectionID: f.ConnectionID,
+				},
+			},
+		}
+		if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+			log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+		}
+	}
+	defer delete(portForwarders, f.SessionID)
+	f.ctxCancel()
+	return f.conn.Close()
+}
+
+func (f *MenderPortForwarder) Read() {
+	errChan := make(chan error)
+	dataChan := make(chan []byte)
+
+	go func() {
+		data := make([]byte, portForwardBuffSize)
+
+		for {
+			n, err := f.conn.Read(data)
+			if err != nil {
+				errChan <- err
+				break
+			}
+			if n > 0 {
+				tmp := make([]byte, n)
+				copy(tmp, data[:n])
+				dataChan <- tmp
+			}
+		}
+	}()
+
+	for {
+		select {
+		case err := <-errChan:
+			if err != io.EOF {
+				log.Errorf("port-forward[%s/%s] error: %v\n", f.SessionID, f.ConnectionID, err.Error())
+			}
+			f.Close(true)
+		case data := <-dataChan:
+			log.Debugf("port-forward[%s/%s] read %d bytes", f.SessionID, f.ConnectionID, len(data))
+
+			// lock the ack mutex, we don't allow more than one in-flight message
+			f.mutexAck.Lock()
+
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypePortForward,
+					MsgType:   wspf.MessageTypePortForward,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						propertyConnectionID: f.ConnectionID,
+					},
+				},
+				Body: data,
+			}
+			if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+				log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+			}
+		case <-time.After(portForwardConnectionTimeout):
+			f.Close(true)
+		case <-f.ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *MenderPortForwarder) Write(body []byte) error {
+	log.Debugf("port-forward[%s/%s] write %d bytes", f.SessionID, f.ConnectionID, len(body))
+	_, err := f.conn.Write(body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func PortForward() Constructor {
+	f := HandlerFunc(portForwardHandler)
```

Comment:
`HandlerFunc` was intended for stateless handlers that doesn't consume external resources. The problem with using it here is that it does not free the local socket & port on the device if the point-to-point connection closes unexpectedly. The `SessionHandler.Close` is called whenever the health check for the `Session` fails, and should be responsible for cleaning up allocated resources. IOW; convert `MenderPortForwarder` to a `SessionHandler` by adding a `Close() error` member and make `portForwardHandler` a member function with the name `ServeProtoMsg(*ws.ProtoMsg, ResponseWriter)`.

Reply:
@alfrunes something like this? https://github.com/mendersoftware/mender-connect/pull/15/commits/0fe7762f407d259b112527f1c57d6824c476a669